### PR TITLE
LLVM_Util: add support for masked operations

### PR DESCRIPTION
**Pull request by lgritz**
Work by Alex Wells.

This will be used extensively by the SIMD batched shading.

* Added support for masked operations by managing a stack of masks
  that were created by conditionals or control flow. Detailed
  explanation of the Masking Strategy in llvm_util.h.

* Added support for logically masked functions (handling of masked
  return from inside OSL functions)

* Added support for masked loops (handling of masked break and or
  masked continue).

* Added support for masked exit (handling of masked exit from entire
  shader).

* Added support for management of when masking is required with a
  ScopedMasking object to allow nesting during code gen.

* op_store (and other store variations) respect the current masking
  requirement and current top of the mask stack (not just the top
  entry, but a combined view of the entire mask stack).

* Added support for results of conditional operations to remain
  boolean.  Most masking operations are expecting vector of bool
  values, not a vector of integers., e.g. when vector width is 16, for
  scalar or <16 x i1> for wide. When a conditional result is used by
  another operation a numerical value it is automatically converted to
  an integer.

* Added `llvm::Type *type_native_mask() const;` because different
  ISA's may have different representations of a mask from llvm's
  vector of bits that comparison operations emit.  And we need varying
  boolean symbols to have the correct data type (size) on the stack
  and in data structures.

* Added gather and scatter operations.

* Added helpers to convert from bool<->int, native_mask<->llvm_mask

Signed-off-by: Larry Gritz <lg@larrygritz.com>

### CODE REVIEW

**Review by AlexMWells — APPROVED** [review]
LGTM

File: src/include/OSL/llvm_util.h [hunk 1]
```diff
@@ -49,7 +50,7 @@ namespace llvm {
 OSL_NAMESPACE_ENTER
 
 namespace pvt {   // OSL::pvt
-
+class ShaderInstance;
```
Comment (fpsunflower) [inline src/include/OSL/llvm_util.h:53]: Is this forward declaration used? I thought `llvm_util.h` was trying to stay mostly agnostic about OSL specifics.

File: src/include/OSL/llvm_util.h [hunk 2]
```diff
@@ -306,6 +309,189 @@ class OSLEXECPUBLIC LLVM_Util {
     /// Are we inside a function?
     bool inside_function() const;
 
+    // Overview of masking strategy
+    //
+    // For if/then based on a conditional, the result of the conditional is
+    // conditionals written to a symbol that is a wide boolean.  To handle
+    // nested we will keep a stack of conditional masks.  When a new
```
Comment (fpsunflower) [inline src/include/OSL/llvm_util.h:316]: typo: `nesting`

File: src/include/OSL/llvm_util.h [hunk 3]
```diff
@@ -306,6 +309,189 @@ class OSLEXECPUBLIC LLVM_Util {
...
+    // conditional scope is exited, we just pop the top of the conditional
+    // mask stack.
+    //
+    // Loops are handled by pushing their conditional result onto the
+    // conditional mask stack, and popping it back off at the end of the
+    // loop.
+    //
+    // Handling of early outs (exit, return, break, continue) is done by
+    // storing a mask representing the lanes which are going to be
+    // deactivated (or continue executing) to an allocated location the
```
Comment (fpsunflower) [inline src/include/OSL/llvm_util.h:329]: typo: `in the stack`

**Review by fpsunflower — APPROVED** [review]
LGTM, just a few minor notes below

**Comment by lgritz** [comment]
Quite right on the typos and other comments, @fpsunflower . Pushed an update and will merge after all the CI tests pass.

File: src/include/OSL/llvm_util.h (hunk 1 above)
Comment (AlexMWells) [reply src/include/OSL/llvm_util.h]: Doesn't appear its used.

**Review by AlexMWells — APPROVED** [review]

**Pull request merged**